packages/@react-spectrum/s2/src/CardView.tsx

```diff
@@ -0,0 +1,513 @@
+    let newLayoutInfos = new Map();
+    let addNode = (key, node) => {
+      let oldLayoutInfo = this.layoutInfos.get(key);
+      let height = itemHeight;
+      let estimatedSize = true;
+      if (oldLayoutInfo) {
+        height = oldLayoutInfo.rect.height;
+        estimatedSize = invalidationContext.sizeChanged || oldLayoutInfo.estimatedSize;
+      }
+
+      // Figure out which column to place the item in, and compute its position.
+      let column = columnHeights.reduce((minIndex, h, i) => h < columnHeights[minIndex] ? i : minIndex, 0);
+      let x = horizontalSpacing + column * (itemWidth + horizontalSpacing);
+      let y = columnHeights[column];
+
+      let rect = new Rect(x, y, itemWidth, height);
+      let layoutInfo = new LayoutInfo(node.type, key, rect);
+      layoutInfo.estimatedSize = estimatedSize;
+      layoutInfo.allowOverflow = true;
+      newLayoutInfos.set(key, layoutInfo);
+
+      columnHeights[column] += layoutInfo.rect.height + this.minSpace.height;
+      return layoutInfo;
+    };
+
+    let skeletonCount = 0;
+    for (let node of this.virtualizer.collection) {
+      if (node.type === 'skeleton') {
+        // Add skeleton cards until every column has at least one, and we fill the viewport.
+        let startingHeights = [...columnHeights];
+        while (
+          !columnHeights.every((h, i) => h !== startingHeights[i]) ||
+          Math.min(...columnHeights) < this.virtualizer.visibleRect.height
+        ) {
+          let layoutInfo = addNode(`${node.key}-${skeletonCount++}`, node);
+          layoutInfo.content = this.layoutInfos.get(layoutInfo.key)?.content || {...node};
+        }
+        break;
+      } else {
+        addNode(node.key, node);
+      }
+    }
+
+    // Reset all columns to the maximum for the next section
+    let maxHeight = Math.max(...columnHeights);
+    this.contentSize = new Size(this.virtualizer.visibleRect.width, maxHeight);
+    this.layoutInfos = newLayoutInfos;
+  }
+
+  getLayoutInfo(key: Key): LayoutInfo {
+    return this.layoutInfos.get(key)!;
+  }
+
+  getContentSize(): Size {
+    return this.contentSize;
+  }
+
+  getVisibleLayoutInfos(rect: Rect): LayoutInfo[] {
+    let layoutInfos: LayoutInfo[] = [];
+    for (let layoutInfo of this.layoutInfos.values()) {
+      if (layoutInfo.rect.intersects(rect) || this.virtualizer.isPersistedKey(layoutInfo.key)) {
+        layoutInfos.push(layoutInfo);
+      }
+    }
+    return layoutInfos;
+  }
+
+  updateItemSize(key: Key, size: Size) {
+    let layoutInfo = this.layoutInfos.get(key);
+    if (!size || !layoutInfo) {
+      return false;
+    }
+
+    if (size.height !== layoutInfo.rect.height) {
+      let newLayoutInfo = layoutInfo.copy();
+      newLayoutInfo.rect.height = size.height;
+      newLayoutInfo.estimatedSize = false;
+      this.layoutInfos.set(key, newLayoutInfo);
+      return true;
+    }
+
+    return false;
+  }
+
+  // Override keyboard navigation to work spacially.
```

Comment:
Technically these overrides work for all grid layouts if you include keyAbove/keyBelow the same way. It's just that in a grid there's only one overlap candidate ever possible.

Any reason not to change how the base version works to this and then not worry about overriding?

Reply:
The spacial navigation algorithm isn't that simple to implement performantly with just the DOM. There's not a good way to query for elements inside a rectangle that I know of